File: frontend/types/company.ts
export type CompanyStage =
  | "unknown"
  | "pre_seed"
  | "seed"
  | "series_a"
  | "series_b"
  | "growth"
  | "public";

export type CompanySource =
  | "manual"
  | "yc"
  | "product_hunt"
  | "hacker_news"
  | "wellfound"
  | "company_website"
  | "rss"
  | "other";

export type Company = {
  id: string;
  name: string;
  website_url: string | null;
  normalized_domain: string;
  description: string | null;
  country: string | null;
  city: string | null;
  stage: CompanyStage;
  source: CompanySource;
  employee_count_min: number | null;
  employee_count_max: number | null;
  founded_year: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string | null;
};

export type CompanyCreateInput = {
  name: string;
  website_url: string;
  description?: string | null;
  country?: string | null;
  city?: string | null;
  stage?: CompanyStage;
  source?: CompanySource;
  employee_count_min?: number | null;
  employee_count_max?: number | null;
  founded_year?: number | null;
  is_active?: boolean;
};

export type CompanyUpdateInput = Partial<CompanyCreateInput>;

export type CompanyCreate = CompanyCreateInput;
export type CompanyUpdate = CompanyUpdateInput;

export type ListCompaniesParams = {
  page?: number;
  page_size?: number;
  search?: string;
  source?: CompanySource;
  stage?: CompanyStage;
  is_active?: boolean;
};

File: backend/app/schemas/company.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.utils.enums import CompanySource, CompanyStage


CURRENT_YEAR = datetime.now().year


class CompanyBase(BaseModel):
    name: str
    website_url: str
    description: str | None = None
    country: str | None = None
    city: str | None = None
    stage: CompanyStage = CompanyStage.UNKNOWN
    source: CompanySource = CompanySource.OTHER
    employee_count_min: int | None = Field(default=None, ge=0)
    employee_count_max: int | None = Field(default=None, ge=0)
    founded_year: int | None = Field(default=None, ge=1800, le=CURRENT_YEAR)
    is_active: bool = True

    @model_validator(mode="after")
    def validate_employee_range(self):
        if (
            self.employee_count_min is not None
            and self.employee_count_max is not None
            and self.employee_count_min > self.employee_count_max
        ):
            raise ValueError("employee_count_min must be <= employee_count_max")
        return self


class CompanyCreate(CompanyBase):
    pass


class CompanyUpdate(BaseModel):
    name: str | None = None
    website_url: str | None = None
    description: str | None = None
    country: str | None = None
    city: str | None = None
    stage: CompanyStage | None = None
    source: CompanySource | None = None
    employee_count_min: int | None = Field(default=None, ge=0)
    employee_count_max: int | None = Field(default=None, ge=0)
    founded_year: int | None = Field(default=None, ge=1800, le=CURRENT_YEAR)
    is_active: bool | None = None

    @model_validator(mode="after")
    def validate_employee_range(self):
        if (
            self.employee_count_min is not None
            and self.employee_count_max is not None
            and self.employee_count_min > self.employee_count_max
        ):
            raise ValueError("employee_count_min must be <= employee_count_max")
        return self


class CompanyRead(CompanyBase):
    model_config = ConfigDict(from_attributes=True)

    id: str
    website_url: str | None = None
    normalized_domain: str
    created_at: datetime
    updated_at: datetime | None = None


class CompanyListItem(CompanyRead):
    pass
